app/controllers/api/account.php

```diff
@@ -1647,6 +1654,99 @@
         throw new Exception('Session not found', 404);
     });
 
+App::patch('/v1/account/sessions/:sessionId/oauth2-tokens')
+    ->desc('Update OAUth2 Tokens')
+    ->groups(['api', 'account'])
+    ->label('scope', 'account')
+    ->label('event', 'account.sessions.update')
+    ->label('sdk.auth', [APP_AUTH_TYPE_SESSION, APP_AUTH_TYPE_JWT])
+    ->label('sdk.namespace', 'account')
+    ->label('sdk.method', 'updateOAuth2Tokens')
+    ->label('sdk.description', '/docs/references/account/update-oauth2-tokens.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_SESSION)
+    ->label('abuse-limit', 10)
+    ->param('sessionId', null, new UID(), 'Session ID. Use the string \'current\' to update the current device session.')
+    ->param('force', false, new Boolean(), 'Should generate new token even if current one is still valid?', true)
+    ->inject('request')
+    ->inject('response')
+    ->inject('user')
+    ->inject('dbForProject')
+    ->inject('project')
+    ->inject('locale')
+    ->inject('audits')
+    ->inject('events')
+    ->inject('usage')
+    ->action(function ($sessionId, $force, $request, $response, $user, $dbForProject, $project, $locale, $audits, $events, $usage) {
+        /** @var Appwrite\Utopia\Request $request */
+        /** @var boolean $force */
+        /** @var Appwrite\Utopia\Response $response */
+        /** @var Utopia\Database\Document $user */
+        /** @var Utopia\Database\Database $dbForProject */
+        /** @var Utopia\Database\Document $project */
+        /** @var Utopia\Locale\Locale $locale */
+        /** @var Appwrite\Event\Event $audits */
+        /** @var Appwrite\Event\Event $events */
+        /** @var Appwrite\Stats\Stats $usage */
+
+        $sessionId = ($sessionId === 'current')
+            ? Auth::sessionVerify($user->getAttribute('sessions'), Auth::$secret)
+            : $sessionId;
+
+        $sessions = $user->getAttribute('sessions', []);
+
+        foreach ($sessions as $key => $session) {/** @var Document $session */
+            if ($sessionId == $session->getId()) {
+                $expireAt = (int) $session->getAttribute('providerAccessTokenExpiry');
+                if(\time() < $expireAt - 5 && !$force) { // 5 seconds time-sync and networking gap, to be safe
+                    return $response->noContent();
```

Comment:
Not sure this is the right response here. We usually use no-content only on delete operations.

Reply:
Added null safety

Reply:
I updated this part by removing the force parameter. If a developer decides to hit this endpoint, we always renew the token. They can check expiryDate from session to prevent firing it too often. Worst-case scenario, their app ID/secret hits the API limit. Not a concern for Appwrite Cloud.